🌱 Improve dry run for topology changes to dry run server side apply

chrischdi opened the pull request:
**What this PR does / why we need it**:

This proposes an improvement on how we implement dry run for detecting changes by actually using dry run server side apply.

The proposed implementation has the following advantages

* It uses the same logic as using the real server side apply patch operation
* The diff logic relies on comparing the current object to the object how it would look like after server side apply
* The diff logic does not rely on how managed fields are used by the kube-apiserver and requires no knowledge about the handled object or used schema.

Disadvantages:

* Using dry run server side apply requires more requests to the kube-apiserver because it requires an additional request for each comparison.

TODOs:
- [x] Add provider implementation documentation:
  - for template resources which get rotated (known types are: `InfrastructureMachineTemplate` or `BootstrapTemplate`): If there are existing validating webhooks which block due to immutability when a resource gets updated (`ValidateUpdate`) they need to implement the `sigs.k8s.io/cluster-aip/util/webhooks.TopologyAwareValidator` instead which provides a bool to decide skipping immutability checks

sbueringer commented on internal/controllers/topology/cluster/structuredmerge/serversidepathhelper.go:
```diff
@@ -78,8 +82,6 @@ func NewServerSidePatchHelper(original, modified client.Object, c client.Client,
 			}
 		}
-
-		filterObject(originalUnstructured, helperOptions)
```
Q: Why do we not need this anymore?

chrischdi replied:
It is getting filtered in `dryRunSSAPatch` and is only here for comparison to the `dryRunUnstructured`

sbueringer commented on internal/controllers/topology/cluster/structuredmerge/serversidepathhelper.go:
```diff
@@ -41,21 +56,10 @@ type serverSidePatchHelper struct {
 
 // NewServerSidePatchHelper returns a new PatchHelper using server side apply.
-func NewServerSidePatchHelper(original, modified client.Object, c client.Client, opts ...HelperOption) (PatchHelper, error) {
+func NewServerSidePatchHelper(ctx context.Context, original, modified client.Object, c client.Client, opts ...HelperOption) (PatchHelper, error) {
```
Please propagate ctx into cleanupLegacyManagedFields

sbueringer commented on internal/controllers/topology/cluster/structuredmerge/serversidepathhelper.go:
```diff
@@ -144,6 +148,57 @@ func (h *serverSidePatchHelper) Patch(ctx context.Context) error {
+
+	err := c.Patch(ctx, dryRunUnstructured, client.Apply, client.DryRunAll, client.FieldOwner(TopologyManagerName), client.ForceOwnership)
+	if err != nil {
+		return false, false, err
```
nit: error wrap

sbueringer replied:
Same below

chrischdi replied:
done :-)

sbueringer reviewed (commented):
Very high level review & ignored the tests for now

Looks good

chrischdi commented:
/hold

competes with #6709

/test all

chrischdi commented:
/test all

chrischdi commented:
e2e currently fails due to:
```
error reconciling the Cluster topology: failed to create patch helper for DockerMachineTemplate/k8s-upgrade-with-runtimesdk-1uasfg-control-plane-52rwq: failed to determine changes via dryRunSSAPatch: failed to request dry-run server side apply: admission webhook "validation.dockermachinetemplate.infrastructure.cluster.x-k8s.io" denied the request: DockerMachineTemplate.infrastructure.cluster.x-k8s.io "k8s-upgrade-with-runtimesdk-1uasfg-control-plane-52rwq" is invalid: spec.template.spec: Invalid value: v1beta1.DockerMachineTemplate{TypeMeta:v1.TypeMeta{Kind:"DockerMachineTemplate", APIVersion:"infrastructure.cluster.x-k8s.io/v1beta1"}, ObjectMeta:v1.ObjectMeta{Name:"k8s-upgrade-with-runtimesdk-1uasfg-control-plane-52rwq", GenerateName:"", Namespace:"k8s-upgrade-with-runtimesdk-t58uv4", SelfLink:"", UID:"a52c48ad-71e8-423b-98f5-5c4f0861479b", ResourceVersion:"2014", Generation:2, CreationTimestamp:time.Date(2022, time.June, 24, 8, 12, 20, 0, time.Local), DeletionTimestamp:<nil>, DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string{"cluster.x-k8s.io/cluster-name":"k8s-upgrade-with-runtimesdk-1uasfg", "topology.cluster.x-k8s.io/owned":""}, Annotations:map[string]string{"cluster.x-k8s.io/cloned-from-groupkind":"DockerMachineTemplate.infrastructure.cluster.x-k8s.io", "cluster.x-k8s.io/cloned-from-name":"quick-start-control-plane"}, OwnerReferences:[]v1.OwnerReference{v1.OwnerReference{APIVersion:"cluster.x-k8s.io/v1beta1", Kind:"Cluster", Name:"k8s-upgrade-with-runtimesdk-1uasfg", UID:"1836653c-8570-43a8-b20b-a677d9f9c689", Controller:(*bool)(nil), BlockOwnerDeletion:(*bool)(nil)}}, Finalizers:[]string(nil), ZZZ_DeprecatedClusterName:"", ManagedFields:[]v1.ManagedFieldsEntry{v1.ManagedFieldsEntry{Manager:"capi-topology", Operation:"Apply", APIVersion:"infrastructure.cluster.x-k8s.io/v1beta1", Time:time.Date(2022, time.June, 24, 8, 18, 27, 0, time.Local), FieldsType:"FieldsV1", FieldsV1:(*v1.FieldsV1)(0x4000d99140), Subresource:""}}}, Spec:v1beta1.DockerMachineTemplateSpec{Template:v1beta1.DockerMachineTemplateResource{ObjectMeta:v1beta1.ObjectMeta{Labels:map[string]string(nil), Annotations:map[string]string(nil)}, Spec:v1beta1.DockerMachineSpec{ProviderID:(*string)(nil), CustomImage:"kindest/node:v1.24.0", PreLoadImages:[]string(nil), ExtraMounts:[]v1beta1.Mount{v1beta1.Mount{ContainerPath:"/var/run/docker.sock", HostPath:"/var/run/docker.sock", Readonly:false}}, Bootstrapped:false}}}}: DockerMachineTemplate spec.template.spec field is immutable. Please create a new resource instead.
```

We have to catch immutability some more...

sbueringer commented:
Hm. That's an interesting issue. Any ideas on how to work around this? A lot of InfrastructureMachineTemplates are immutable

sbueringer commented:
> > Member
> 
> Yes, catch the error (what we already do) but detect it as `return true, true, nil` in this case

I probably misunderstood. My impression is we can't run SSA on immutable templates because the Validation Webhook will block it so then we won't get a meaningful diff?

Does the SSA give us a result despite the validation webhook blocking the request?

chrischdi commented:
> > > Member
> > 
> > 
> > Yes, catch the error (what we already do) but detect it as `return true, true, nil` in this case
> 
> I probably misunderstood. My impression is we can't run SSA dryrun on immutable templates because the Validation Webhook will block it so then we won't get a meaningful diff?
> 
> Does the SSA dryrun give us a result despite the validation webhook blocking the request?

See the implementation. Thats imho the best we can do here: identify if the error causes are **only** of spec fields which got are marked as invalid.

sbueringer commented:
> > > > Member
> > > 
> > > 
> > > Yes, catch the error (what we already do) but detect it as `return true, true, nil` in this case
> > 
> > 
> > I probably misunderstood. My impression is we can't run SSA dryrun on immutable templates because the Validation Webhook will block it so then we won't get a meaningful diff?
> > Does the SSA dryrun give us a result despite the validation webhook blocking the request?
> 
> See the implementation. Thats imho the best we can do here: identify if the error causes are **only** of spec fields which got are marked as invalid.

Just that I interpret the code correctly. If we get an CauseTypeFieldValueInvalid error from a field in spec. we assume we have changes?

So as long as immutable errors are returned as CauseTypeFieldValueInvalid and with the correct path we detect immutable errors as changes. Which is correct because without changes no immutable error.

I think this doesn't cover cases where e.g. the CAPA provider returns a BadRequest error? [CauseTypeFieldValueInvalid](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/api/v1beta1/awsmachinetemplate_webhook.go#L141-L143)

chrischdi commented:
/test all

chrischdi commented:
Still WIP.


Updated the branch to:

* cherry-pick #6742 
* cherry-pick #6738 
* implemented a custom validator interface
* did adjust the affected validators to the new interface to be topology aware

chrischdi commented:
/test all

chrischdi commented:
/test all

sbueringer commented on controlplane/kubeadm/api/v1beta1/kubeadmcontrolplanetemplate_webhook.go:
```diff
@@ -27,12 +27,19 @@ import (
 
 const kubeadmControlPlaneTemplateImmutableMsg = "KubeadmControlPlaneTemplate spec.template.spec field is immutable. Please create new resource instead."
 
 func (r *KubeadmControlPlaneTemplate) SetupWebhookWithManager(mgr ctrl.Manager) error {
```
I think we only have to do this for templates that we rotate. For regular objects it's not relevant as the dryrun SSA call is identical to what we do later on. So any validation error aligns with the one we would otherwise get with SSA later.

Quick overview:
* no rotation
    * KubeadmControlPlaneTemplate (ControlPlaneTemplate) is reconciled to KubeadmControlPlane
    * DockerClusterTemplate (InfrastructureClusterTemplate) is reconciled to DockerCluster
* rotation
    * DockerMachineTemplate (InfrastructureMachineTemplate) is reconciled to DockerMachineTemplate
    * KubeadmConfigTemplate (BootstrapConfigTemplate) is reconciled to KubeadmConfigTemplate

=> As KubeadmConfigTemplate doesn't have immutability checks we only have to do it for DockerMachineTemplate

chrischdi commented on exp/admission/topology_validator.go:
```diff
@@ -0,0 +1,178 @@
+
+const (
+	// TopologyDryRunAnnotation foo.
+	TopologyDryRunAnnotation = "topology.cluster.x-k8s.io/dry-run"
```
it is not runtime sdk, also we cannot add it to clusterv1 because of circular dependency stuff.

At best we even keep it somewhere in internal?

sbueringer replied:
Very good point with runtime, ups 2nd time today ... :) 

How does adding it to clusterv1 create a circular dependency?

If I see correctly only serversideapplypathhelper and topology_validator need it

sbueringer commented on internal/controllers/topology/cluster/structuredmerge/serversidepathhelper.go:
```diff
@@ -96,12 +100,15 @@ func NewServerSidePatchHelper(original, modified client.Object, c client.Client,
+		var err error
+		hasChanges, hasSpecChanges, err = dryRunSSAPatch(ctx, c,
+			originalUnstructured,
+			modifiedUnstructured,
```
Can we also use an input struct and handover `dryRunUnstructured: modifiedUnstructured.DeepCopy`

Just to make 100% clear that modifiedUnstructured cannot be modified inside of this func?

chrischdi replied:
Its now kind of mixed. Passing `ctx` in a struct results in a linting error. Because of that I excluded ctx from the struct.

sbueringer commented on internal/controllers/topology/cluster/structuredmerge/serversidepathhelper.go:
```diff
@@ -140,10 +147,76 @@ func (h *serverSidePatchHelper) Patch(ctx context.Context) error {
+	}
+
+	// Drop the other fields which are not part of our intent.
+	filterObject(dryRunUnstructured, dryRunHelperOptions)
+	filterObject(originalUnstructured, dryRunHelperOptions)
```
Q: Is there a difference between doing this here vs one level above? 
(dryRunUnstructured is a copy of modifiedUnstructured so we already ran filterObject on it, and if we can keep the filterObject of originalUnstructured one level above it feels a bit easier to reason about what's going on)

Of course if the change in allowedPaths is relevant here we can't do that

chrischdi replied:
Yes, so no we can't move it up.

For dryRunUnstructured: it gets filled via the client.Patch request.

For originalUnstructured: inside this function the helper preserves the managedFields which are important for detecting ownership changes.

chrischdi commented on test/infrastructure/docker/api/v1beta1/dockermachinetemplate_webhook.go:
```diff
@@ -24,35 +24,44 @@ import (
 	}
-	if !reflect.DeepEqual(m.Spec.Template.Spec, old.Spec.Template.Spec) {
-		allErrs = append(allErrs, field.Invalid(field.NewPath("spec", "template", "spec"), m, dockerMachineTemplateImmutableMsg))
+	if !skipImmutability {
+		if !reflect.DeepEqual(m.Spec.Template.Spec, old.Spec.Template.Spec) {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("spec", "template", "spec"), m, dockerMachineTemplateImmutableMsg))
+		}
 	}
```
adjusted to a single line if, I'd rather not do the `if skipImmutability {return nil}` to prevent forgetting to validation which would be later on in the func.

chrischdi commented:
Rebased on top of #6785 (because this change is required for the implementation to work)

/retitle 🌱 Improve dry run for topology changes to dry run server side apply
/test all

chrischdi commented:
/test help

chrischdi commented:
/test pull-cluster-api-e2e-full-main
/test pull-cluster-api-e2e-informing-ipv6-main
/test pull-cluster-api-e2e-informing-main
/test pull-cluster-api-e2e-workload-upgrade-1-21-1-22-main
/test pull-cluster-api-e2e-workload-upgrade-1-24-latest-main

sbueringer commented:
> /test pull-cluster-api-e2e-workload-upgrade-1-24-latest-main
For folks who don't know, this one is expected to fail atm.

sbueringer commented on test/infrastructure/docker/api/v1beta1/dockermachinetemplate_webhook.go:
```diff
@@ -17,48 +17,50 @@ limitations under the License.
 
 // ValidateCreate implements webhook.Validator so a webhook will be registered for the type.
-func (m *DockerMachineTemplate) ValidateCreate() error {
+func (*DockerMachineTemplate) ValidateCreate(ctx context.Context, _ runtime.Object) error {
```
Can we intentionally make this a separate struct? (cf. my PR on your fork) In the other cases we use the CustomValidator we use a separate struct for it because we have to, but I think here it would be good to avoid confusion, for consistency with other CustomValidator webhooks and to avoid additional methods on the API type

chrischdi commented on util/defaulting/defaulting.go:
```diff
@@ -60,3 +68,34 @@ func DefaultValidateTest(object DefaultingValidator) func(*testing.T) {
+// DefaultCustomValidateTest returns a new testing function to be used in tests to
+// make sure defaulting webhooks also pass custom validation tests on create,
+// update and delete.
+func DefaultCustomValidateTest(object DefaultingCustomValidator) func(*testing.T) {
```
also currently not used so I dropped it.

chrischdi commented:
/test pull-cluster-api-e2e-full-main
/test pull-cluster-api-e2e-informing-ipv6-main
/test pull-cluster-api-e2e-informing-main
/test pull-cluster-api-e2e-workload-upgrade-1-21-1-22-main

chrischdi commented:
/test pull-cluster-api-e2e-full-main
/test pull-cluster-api-e2e-informing-ipv6-main
/test pull-cluster-api-e2e-informing-main
/test pull-cluster-api-e2e-workload-upgrade-1-21-1-22-main

fabriziopandini commented on util/topology/topology.go:
```diff
@@ -0,0 +1,51 @@
+)
+
+// ShouldSkipImmutabilityChecks returns true if it is a dry-run request and the object has the
+// TopologyDryRunAnnotation annotation set, false otherwise.
```
nit (explaining why we need this)
```suggestion
// TopologyDryRunAnnotation annotation set, false otherwise.
// This ensures that the immutability check is skipped only when dry-running and when the operations has been invoked by the topology controller.
// Instead, kubectl dry-run behavior remains consistent with the one user gets when doing kubectl apply (immutability is enforced).
```

sbueringer commented on util/topology/topology.go:
```diff
@@ -0,0 +1,51 @@
+// ShouldSkipImmutabilityChecks returns true if it is a dry-run request and the object has the
+// TopologyDryRunAnnotation annotation set, false otherwise.
+func ShouldSkipImmutabilityChecks(ctx context.Context, obj metav1.Object) bool {
+	req, err := admission.RequestFromContext(ctx)
+	if err != nil {
+		return false
+	}
```
I'm not sure if this should be inside of this func
1. Calling it inside of a CustomValidator instead guarantees that the error does not occur (given a minimum CR version)
2. This locks the util func itself to a specific (minimum) controller-runtime version. While keeping it outside of it only couples it to the admission API
3. The function signature looks too magical to me. Feels way more explicit to pass in the admission.Request
    * This would also allow folks to use this func in a relatively normal way by implementing the webhook directly with the lower-level CR API (admission.Handler)

Of course the downside would be that folks have to add a few more lines to their webhook


cc @vincepri WDYT?

(but no showstopper from my side)

chrischdi replied:
I'd be okay with both :-). Will move it up for now.

sbueringer commented on test/infrastructure/docker/api/v1beta1/dockermachinetemplate_webhook_test.go:
```diff
@@ -51,10 +60,22 @@ func TestDockerMachineTemplateInvalid(t *testing.T) {
 			wantError:   true,
 		},
+		{
+			name:        "skip immutability check",
```
nit: Should we also add test case for only dry-run and only the annotation being set?

sbueringer commented on test/infrastructure/docker/api/v1beta1/dockermachinetemplate_webhook_test.go:
```diff
@@ -32,11 +38,14 @@ func TestDockerMachineTemplateInvalid(t *testing.T) {
 
 	newTemplate := oldTemplate.DeepCopy()
 	newTemplate.Spec.Template.Spec.ExtraMounts = append(newTemplate.Spec.Template.Spec.ExtraMounts, []Mount{{ContainerPath: "/var/run/docker.sock", HostPath: "/var/run/docker.sock"}}...)
+	newTemplateSkipImmutability := oldTemplate.DeepCopy()
```
```suggestion
	newTemplateSkipImmutability := newTemplate.DeepCopy()
```

I think this only actually proves we skipped the test if the spec of newTemplateSkipImmutability is different than oldTemplate which it currently isn't

sbueringer commented on test/infrastructure/docker/api/v1beta1/dockermachinetemplate_webhook_test.go:
```diff
@@ -51,10 +60,22 @@ func TestDockerMachineTemplateInvalid(t *testing.T) {
 		t.Run(tt.name, func(t *testing.T) {
-			err := tt.newTemplate.ValidateUpdate(tt.oldTemplate)
+			wh := &DockerMachineTemplateWebhook{}
+			ctx := context.TODO()
```
Let's please use context.Background() in unit tests.

context.TODO() is supposed to be a TODO in the sense of that eventually a real context should be propagated to the place where context.TODO() is used and then replace the TODO

sbueringer reviewed (commented):
Reviewed everything except structuredmerge pkg & docs

sbueringer commented on docs/book/src/developer/providers/v1.1-to-v1.2.md:
````diff
@@ -55,20 +55,82 @@ in ClusterAPI are kept in sync with the versions used by `sigs.k8s.io/controller
+  ```
+
+- [Server Side Apply](https://kubernetes.io/docs/reference/using-api/server-side-apply/) implementation in ClusterClass and managed topologies 
+  requires to dry-run changes on templates, and it is required to implement following changes in order to prevent dry-run
+  to return errors due to immutability checks implemented in the web hooks.
````
```suggestion
  requires to dry-run changes on templates. If infrastructure or bootstrap providers have implemented immutability checks in their 
  InfrastructureMachineTemplate or BootstrapConfigTemplate webhooks, it is required to implement the following changes in order to prevent dry-run
  to return errors.
```

I think we have to add when it is required. Only a very small subset of infra providers have implemented the immutability check and it isn't documented / standardized anywhere as far as I can tell.

Q: Should we have a follow-up to document this somewhere in the contract? I think we can't expect folks to read through previous migration docs.

chrischdi replied:
I've adjusted the sentence but will leave this unresolved due to docs.

I'm +1 adding thsi to the contract.

sbueringer replied:
I think we can link to there but in the contract doc it's probably framed slightly different (not as a diff, but rather as what to do in ValidateUpdate if you want to do immutability checks)

chrischdi replied:
Maybe add a sentence like this:

```

Validating webhooks implemented for InfrastructureMachineTemplate or BootstrapConfigTemplate are required to not block due to immutability checks when the controller for managed topology and ClusterClass does
[Server Side Apply](https://kubernetes.io/docs/reference/using-api/server-side-apply/) dry-run requests.
```

to https://main.cluster-api.sigs.k8s.io/developer/providers/webhooks.html#validating-webhooks ? wdyt?

sbueringer replied:
That definitely doesn't hurt. I wonder if we should add a code snippet /example for ValidateUpdate + if it should be here instead: https://main.cluster-api.sigs.k8s.io/developer/providers/contracts.html

sbueringer replied:
With some restriction that it's only required if ClusterClass should be supported (similar to what we have here for using versions, using machines, ..): https://main.cluster-api.sigs.k8s.io/developer/architecture/controllers/control-plane.html#required-spec-fields-for-implementations-using-version

Probably makes sense to hear Fabrizio's take on it

sbueringer commented on docs/book/src/developer/providers/v1.1-to-v1.2.md:
```diff
@@ -55,20 +55,82 @@ in ClusterAPI are kept in sync with the versions used by `sigs.k8s.io/controller
+
+NOTES:
+
+- We are introducing a `DockerMachineTemplateWebhook` struct because we are going to use a controller runtime
+  `CustomValidator` that will allow to skip the immutability check only when the topology controller is dry running
+  while preserving the validation behaviour for all the other cases.
```
```suggestion
- We are introducing a `DockerMachineTemplateWebhook` struct because we are going to use a controller runtime
  `CustomValidator`. This will allow to skip the immutability check only when the topology controller is dry running
  while preserving the validation behaviour for all other cases.
```
nit

sbueringer commented on docs/book/src/developer/providers/v1.1-to-v1.2.md:
```diff
@@ -55,20 +55,82 @@ in ClusterAPI are kept in sync with the versions used by `sigs.k8s.io/controller
+- We are introducing a `DockerMachineTemplateWebhook` struct because we are going to use a controller runtime
+  `CustomValidator` that will allow to skip the immutability check only when the topology controller is dry running
+  while preserving the validation behaviour for all the other cases.
+- By using CustomValidators it will be possible to move webhooks to other packages, thus removing controller
+  runtime dependency from the API types. However, choosing to do so or not is up to the providers implementers. 
```
```suggestion
- By using `CustomValidators` it is possible to move webhooks to other packages, thus removing some controller
  runtime dependency from the API types. However, choosing to do so or not is up to the provider implementers
  and independent of this change. 
```
nit

sbueringer commented on internal/controllers/topology/cluster/structuredmerge/dryrun.go:
```diff
@@ -18,285 +18,149 @@ package structuredmerge
+	client client.Client
+	// originalUnstructured contains the current state of the object
+	originalUnstructured *unstructured.Unstructured
+	// dryRunUnstructured contains the intended changes to the object and will used to
```
```suggestion
	// dryRunUnstructured contains the intended changes to the object and will be used to
```
nit

sbueringer commented on internal/controllers/topology/cluster/structuredmerge/dryrun_test.go:
```diff
@@ -18,871 +18,149 @@ package structuredmerge
 	}
 }
+
+type cleanupTopologyDryRunBuilder struct {
```
wdyt about calling this simply `objectBuilder`?

sbueringer commented:
Flake
/retest

sbueringer commented on docs/book/src/developer/providers/v1.1-to-v1.2.md:
````diff
@@ -55,20 +55,83 @@ in ClusterAPI are kept in sync with the versions used by `sigs.k8s.io/controller
+  Subnets Subnets `json:"subnets,omitempty"
+  ```
+
+- [Server Side Apply](https://kubernetes.io/docs/reference/using-api/server-side-apply/) implementation in ClusterClass and managed topologies 
````
Should we add an info that controller-runtime >= v0.12.3 is required?

fabriziopandini replied:
It could be useful to add controller runtime to the list on top of the page as well

sbueringer commented:
lgtm pending squash + https://github.com/kubernetes-sigs/cluster-api/pull/6710#discussion_r914851610

I would be fine with moving https://github.com/kubernetes-sigs/cluster-api/pull/6710#discussion_r914851610 to a follow-up issue so we get this PR merged ASAP and still get a few days time in CI.

chrischdi commented:
> lgtm pending squash + [#6710 (comment)](https://github.com/kubernetes-sigs/cluster-api/pull/6710#discussion_r914851610)
> 
> I would be fine with moving [#6710 (comment)](https://github.com/kubernetes-sigs/cluster-api/pull/6710#discussion_r914851610) to a follow-up issue so we get this PR merged ASAP and still get a few days time in CI.

Regarding squash: @fabriziopandini do we still want to keep the CAPD change as seperate commit?

Sounds good to move it to a follow-up. I think we could even link the commit for the CAPD change if we want (which we can do after merge only)?!

chrischdi commented:
/test pull-cluster-api-e2e-full-main
/test pull-cluster-api-e2e-informing-ipv6-main
/test pull-cluster-api-e2e-informing-main
/test pull-cluster-api-e2e-workload-upgrade-1-21-1-22-main

chrischdi commented:
Runtime SDK flake: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api/6710/pull-cluster-api-e2e-full-main/1544943529262845952

Getting addressed by #6817 ? @killianmuldoon 

/retest

fabriziopandini commented on util/topology/topology.go:
```diff
@@ -0,0 +1,46 @@
+// TopologyDryRunAnnotation annotation set, false otherwise.
+// This ensures that the immutability check is skipped only when dry-running and when the operations has been invoked by the topology controller.
+// Instead, kubectl dry-run behavior remains consistent with the one user gets when doing kubectl apply (immutability is enforced).
+func ShouldSkipImmutabilityChecks(req admission.Request, obj metav1.Object) bool {
```
Can we add a test for this func?

chrischdi commented:
/cherry-pick release-1.2

chrischdi commented:
/test pull-cluster-api-e2e-full-main
/test pull-cluster-api-e2e-informing-ipv6-main
/test pull-cluster-api-e2e-informing-main
/test pull-cluster-api-e2e-workload-upgrade-1-21-1-22-main

sbueringer commented:
Thx
/lgtm

sbueringer commented:
/hold cancel
